modules/ravestate_telegramio/telegram_bot.py

```diff
@@ -116,42 +134,125 @@ def handle_photo(bot: Bot, update: Update):
+    is_master_process = child_conn is None
+    if is_master_process:
+        # Master Process -> Start the bot
+        # Create the EventHandler and pass it your bots token.
+        token = ctx.conf(key=TOKEN_CONFIG_KEY)
+        if not token:
+            logger.error(f'{TOKEN_CONFIG_KEY} is not set. Shutting down telegramio')
+            return Delete()
+        child_config_paths_list = ctx.conf(key=CHILD_FILES_CONFIG_KEY)
+        if not ctx.conf(key=ALL_IN_ONE_CONTEXT_CONFIG_KEY) and (
+                not child_config_paths_list or not isinstance(child_config_paths_list, list)
+                or not all(os.path.isfile(child_config_path) for child_config_path in child_config_paths_list)):
+            logger.error(f'{CHILD_FILES_CONFIG_KEY} is not set (correctly). Shutting down telegramio')
+            return Delete()
+
+        updater: Updater = Updater(token)
+        # Get the dispatcher to register handlers
+        dp: Dispatcher = updater.dispatcher
+        if ctx.conf(key=ALL_IN_ONE_CONTEXT_CONFIG_KEY):
+            # handle noncommand-messages with the matching handler
+            dp.add_handler(MessageHandler(Filters.text, handle_text))
+            dp.add_handler(MessageHandler(Filters.photo, handle_photo))
+        else:
+            dp.add_handler(MessageHandler(Filters.text | Filters.photo, handle_input_multiprocess))
+        # log all errors
+        dp.add_error_handler(error)
+        # Start the Bot
+        updater.start_polling()  # non blocking
+
+        if not ctx.conf(key=ALL_IN_ONE_CONTEXT_CONFIG_KEY):
+            while not ctx.shutting_down():
+                # wait for children to write to Pipe and then send message to chat
+                tick_interval = 1. / ctx.conf(mod=Context.core_module_name, key=Context.tick_rate_config)
+                time.sleep(tick_interval)
+                for chat_id, parent_pipe in active_chats.items():
+                    if parent_pipe.poll():
+                        msg = parent_pipe.recv()
+                        if isinstance(msg, str):
+                            updater.bot.send_message(chat_id=chat_id, text=msg)
+                        else:
+                            logger.error(f'Tried sending non-str object as telegram message: {str(msg)}')
+
+    else:
+        # Child Process -> check pipe
+        try:
+            while not ctx.shutting_down():
+                # receive Bot,Update for telegram chat
+                bot, update = child_conn.recv()  # blocking
+                if update.effective_message.photo:
+                    handle_photo(bot, update)
+                elif update.effective_message.text:
+                    handle_text(bot, update)
+                else:
+                    logger.error(f"{MODULE_NAME} recieved an update it cannot handle.")
+        except EOFError:
+            # Pipe was closed -> Parent was killed
+            logger.info("Parent process was killed, therefore the telegram-child will shut down.")
```

Comment:
Would use warning/error level since it is somehow unwanted situation

Reply:
This is info-level because this is the "normal shutdown path".
[Connection Documentation](https://docs.python.org/3.6/library/multiprocessing.html?#multiprocessing.connection.Connection) mentions that EOFError is raised if the other end of the pipe is closed.